### PR TITLE
Replace test infrastructure from umock-c to cmocka.

### PR DESCRIPTION
**Pull request by mamokarz**

### CODE REVIEW

**Review by smcd253 — APPROVED** [review]
lgtm

File: .gitmodules [hunk 1]
```diff
@@ -10,3 +10,6 @@
 [submodule "deps/azure-macro-utils-c"]
```
Comment (danewalton) [inline .gitmodules:10]: are any of these other submodules needed anymore?

Comment (danewalton) [reply .gitmodules:10]: look to remove them?

File: README.md [hunk 2]
```diff
@@ -11,68 +11,156 @@ This repository contains the following:
...
+### CMake
+
+1. Install the required prerequisites:
+   - [CMake](https://cmake.org/download/) version 3.10 or later
+   - C compiler: [MSVC](https://visualstudio.microsoft.com/downloads/#build-tools-for-visual-studio-2019), [gcc](https://gcc.gnu.org/) or [clang](https://clang.llvm.org/) are recommended
+   - [git](https://git-scm.com/downloads) to clone our Azure uLib repository with the desired tag
+
+2. Clone our Azure uLib repository, optionally using the desired version tag.
+
+        git clone https://github.com/Azure/azure-ulib-c
```
Comment (danewalton) [inline README.md:78]:
```suggestion
        git clone https://github.com/Azure/azure-ulib-c --recursive
```

File: deps/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  ${CMAKE_CURRENT_BINARY_DIR}
+  $<BUILD_INTERFACE:${CMOCKA_INCLUDE_DIR}>
+)
+
+target_compile_options(az_cmocka PRIVATE -DHAVE_CONFIG_H)
+
+if ( MSVC )
+  target_compile_options(az_cmocka PRIVATE /wd4267 /wd4244 /wd4267 /wd4996)
+else ()
+  target_compile_options(az_cmocka PRIVATE -Wno-implicit-function-declaration -Wno-int-conversion)
```
Comment (danewalton) [inline deps/CMakeLists.txt:32]: are these consistent across GCC and clang?

**Review by danewalton — APPROVED** [review]

File: .gitmodules (hunk 1 above)
Comment (mamokarz) [reply .gitmodules:10]: I had removed that, I don't know why it is here. I'll remove again. 

---
In reply to: [597935610](https://github.com/Azure/azure-ulib-c/pull/62#discussion_r597935610) [](ancestors = 597935610)

File: CMakeLists.txt [hunk 4]
```diff
@@ -76,18 +84,20 @@ target_link_libraries(azure_ulib_c
...
 
 ulib_set_target_build_properties(azure_ulib_c)
 
-if (${BUILD_TESTING})
+# default for Unit testing with cmocka is OFF, however, this will be ON on CI and tests must
+# pass before committing changes
+if (UNIT_TESTING)
+    add_subdirectory(${CMAKE_CURRENT_LIST_DIR}/deps)
+
+    set(TEST_DIRECTORY ${PROJECT_SOURCE_DIR}/tests)
```
[the remark this answers is not in the document]
Comment (mamokarz) [reply CMakeLists.txt:96]: No, I'll fix it.

---
In reply to: [597781671](https://github.com/Azure/azure-ulib-c/pull/62#discussion_r597781671) [](ancestors = 597781671)

File: samples/ipc_hardware_update/consumers/my_consumer.c [hunk 5]
```diff
@@ -8,11 +8,16 @@
...
+static const char hello[] = "Hello world! This is a test to display a message.";
+static const int32_t hello_size = sizeof(hello) - 1;
+static const char bunny_1[] = "(\\(\\";
+static const int32_t bunny_1_size = sizeof(bunny_1) - 1;
+static const char bunny_2[] = "( -.-)";
+static const int32_t bunny_2_size = sizeof(bunny_2) - 1;
+static const char bunny_3[] = "o_(\")(\")";
+static const int32_t bunny_3_size = sizeof(bunny_3) - 1;
+static const char bunny_11[] = "/)/)";
+static const int32_t bunny_11_size = sizeof(bunny_11) - 1;
```
[the remark this answers is not in the document]
Comment (mamokarz) [reply samples/ipc_hardware_update/consumers/my_consumer.c:20]: :D

---
In reply to: [597783429](https://github.com/Azure/azure-ulib-c/pull/62#discussion_r597783429) [](ancestors = 597783429)

File: deps/CMakeLists.txt (hunk 3 above)
Comment (mamokarz) [reply deps/CMakeLists.txt:32]: Yep

---
In reply to: [597938256](https://github.com/Azure/azure-ulib-c/pull/62#discussion_r597938256) [](ancestors = 597938256)

**Pull request merged**